### PR TITLE
Mwfq

### PR DESCRIPTION
**Pull request by sveinbjornt**
A hodgepodge (or a veritable smorgasbord, if you will) of improvements throughout the Greynir codebase. This really should have been done on different branches with separate pull requests but here we are:

* Various improvements in word frequency page (broken down by week for longer periods, new and improved input word lookup and parsing, entities and persons can now be queried, periods can be clicked to see articles, etc.)
* Added leap year query handling to date query module, refactored entire module (still buggy)
* Stats on people, places pages can now be requested by month, in addition to week and day (plenty fast)
* Many additions to Vocab.conf
* Some additions in geo.py and geo data
* Now using new and improved iceaddr module for placename lookup
* Utility script orn.py to inspect our handling of Icelandic placenames ("örnefni")
* Various improvements in gencorp.py script
* Additional tests, mostly for date query module
* Refactoring and renaming in templates
* Fonts now loaded locally instead of via Google Fonts (prevents tracking, slightly increases our bandwidth usage). Also dropped custom font support for legacy browsers, only providing them in woff/woff2 formats (ttf fonts are huge).
* About page now shows server platform name (e.g. Linux, Darwin, Windows) in addition to Python and ReynirPackage versions
* Deployment script adds git commit hash to about page so we always know exactly which version is currently deployed

### CODE REVIEW

File: config/Vocab.conf [hunk 1]
```diff
@@ -919,6 +958,27 @@ fræsing        fræsinguna       kvk alm ÞFETgr
...
 fræsing        fræsingarinnar   kvk alm EFETgr
 
+# 'lúddíti' vantar í BÍN
+lúddíti        lúddíti            kk alm NFET
+lúddíti        lúddíta            kk alm ÞFET
+lúddíti        lúddíta            kk alm ÞGFET
+lúddíti        lúddíta            kk alm EFET
+
+lúddíti        lúddítinn          kk alm NFETgr
+lúddíti        lúddítan           kk alm ÞFETgr
```
Comment (vthorsteinsson) [inline config/Vocab.conf:968]: lúddítann - vantar n

File: routes/stats.py [hunk 2]
```diff
@@ -46,27 +46,25 @@
 _DEFAULT_STATS_PERIOD = 10
 _MAX_STATS_PERIOD = 30
 _TOP_AUTHORS_PERIOD = 30
-_QUERY_STATS_PERIOD = 10
+
+# TODO: This should be put in a column in the roots table
```
Comment (vthorsteinsson) [inline routes/stats.py:50]: Indeed

File: routes/words.py [hunk 3]
```diff
@@ -31,30 +30,94 @@
...
+
+@routes.route("/words_trends")
+def words_trends():
+    """ Handler for word trends page. """
+    return render_template("words/trends.html", title="Orð")
+
+
+# Word categories permitted in word frequency search
+_VALID_WCATS = frozenset(
+    ("kk", "kvk", "hk", "lo", "so", "person_kk", "person_kvk", "entity", "??")
```
Comment (vthorsteinsson) [inline routes/words.py:57]: Add "person_hk" - I think that is a possibility in the current code and will be seen more in the future

File: routes/words.py [hunk 4]
```diff
@@ -31,30 +30,94 @@
...
+# Human-readable descriptions of word categories
+CAT_UNKNOWN = "??"
+CAT_DESC = {
+    "kk": "kk. no.",
+    "kvk": "kvk. no.",
+    "hk": "hk. no.",
+    "lo": "lo.",
+    "so": "so.",
+    "person_kk": "kk. manneskja",
+    "person_kvk": "kvk. manneskja",
```
Comment (vthorsteinsson) [inline routes/words.py:69]: Add `"person_hk": "hk. manneskja"`

File: routes/words.py [hunk 5]
```diff
@@ -31,30 +30,94 @@
...
 
+def _str2words(wstr, separate_on_whitespace=False):
+    """ Parse string of the form 'w1:cat1, w2:cat2, ...' into a list
+        of word/category tuples. """
+    if not wstr:
+        return None
+    ws = wstr.strip()
+    if not ws:
+        return None
+    words = [w.strip() for w in ws.split(",")]
```
Comment (vthorsteinsson) [inline routes/words.py:97]: Perhaps `[w.strip() for w in ws.split(",") if w.strip()]` for added safety (the case of multiple sequential commas)

File: routes/words.py [hunk 6]
```diff
@@ -31,30 +30,94 @@
...
+        words = " ".join(words).split()
+    return [
+        w.split(":") if len(w.split(":")) == 2 else (w, CAT_UNKNOWN) for w in words
+    ][:_MAX_NUM_WORDS]
+
+
+def _words2str(words):
+    """ Create comma-separated string from (word,cat) tuple list,
+        e.g. "[(a,b),(c,d)] -> "a:b, c:d". """
+    return ", ".join([":".join(w[:2]) if len(w) >= 2 else (w, None) for w in words])
```
Comment (vthorsteinsson) [inline routes/words.py:108]: Should probably be `", ".join([":".join(w[:2]) if len(w) >= 2 else w[0] for w in words])`

File: routes/words.py [hunk 7]
```diff
@@ -64,74 +127,170 @@ def wordfreq():
...
+    # Create word/cat pair from token
+    def cat4token(t):
+        # TODO: Use ReynirPackage lemma lookup function for this
+        w = t.txt
+        if t.kind == TOK.WORD:
+            val = list(filter(lambda m: m.stofn == m.ordmynd, t.val)) or t.val
+            cat = val[0].ordfl if len(val) else CAT_UNKNOWN
+            w = val[0].stofn if len(val) else t.txt
+        elif t.kind == TOK.PERSON:
+            cat = "person_" + t.val[0].gender
```
Comment (vthorsteinsson) [inline routes/words.py:145]: Yup, this can yield `person_hk`

File: routes/words.py [hunk 8]
```diff
@@ -64,74 +127,170 @@ def wordfreq():
...
-
-        # Get word category (ordfl) for each word, if needed
-        valid_cats = ["kk", "kvk", "hk", "lo", "so"]
-        for i, w in enumerate(words):
-            if len(w) < 2 or w[1] not in valid_cats:
-                words[i] = (w[0], cat4word(w[0]))
+    # Create word/cat pair from token
+    def cat4token(t):
+        # TODO: Use ReynirPackage lemma lookup function for this
+        w = t.txt
```
Comment (vthorsteinsson) [inline routes/words.py:139]: `assert t.kind in (TOK.WORD, TOK.PERSON, TOK.ENTITY)` ?

File: scraper.py [hunk 9]
```diff
@@ -571,7 +571,7 @@ def parse_int(i):
                 reparse = True
             elif o in ("-l", "--limit"):
                 # Maximum number of articles to parse
-                limit = parse_int(a)
+                limit = parse_int(a) or limit
```
Comment (vthorsteinsson) [inline scraper.py:574]: Was the intent to use the `fallback` parameter here?

File: templates/locations/icemap.html [hunk 10]
```diff
@@ -49,13 +49,19 @@
          <div class="btn-group pull-right">
             <span class="hidden-xs">
                <button class="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
-                  {{ "Undanfarna viku" if period == 'week' else "Síðasta sólarhring" }} 
+                  {% if period == 'week' %}
+                     Undanfarna viku
+                  {% elif period == 'month' %}
+                     Síðastliðinn mánuð
```
Comment (vthorsteinsson) [inline templates/locations/icemap.html:55]: Strictly speaking it's the last 30 days, no?

File: templates/words/details.html [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+         {% for a in w["articles"] %}
+            <span class="art-link"><a href="{{ url_for('routes.page', id=a.id) | safe }}" 
+               onClick="event.stopPropagation();" data-toggle="tooltip" data-placement="bottom" 
+               title="{{ a.heading }}" rel="nofollow"><img 
+               src="{{ url_for('static', filename = 'sources/' + a.domain + '.png') }}" 
+               width="16" height="16" alt="{{ a.heading }}"/></a></span>
+        {% endfor %}
+      </div>
+   </div>
+{% endfor %}
```
Comment (vthorsteinsson) [inline templates/words/details.html:22]: Add NL at end of file

File: templates/words_details.html [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+               onClick="event.stopPropagation();" data-toggle="tooltip" data-placement="bottom" 
+               title="{{ a.heading }}" rel="nofollow"><img 
+               src="{{ url_for('static', filename = 'sources/' + a.domain + '.png') }}" 
+               width="16" height="16" alt="{{ a.heading }}"/></a></span>
+        {% endfor %}
+      </div>
+   </div>
+{% endfor %}
+
+
```
Comment (vthorsteinsson) [inline templates/words_details.html:21]: Cut NL at end of file

File: tree.py [hunk 13]
```diff
@@ -774,6 +774,16 @@ def contained_date(self):
             self._aux = json.loads(self.aux)
         return self._aux[0], self._aux[1], self._aux[2]
 
+    @property
+    def contained_year(self):
+        """ Return a year from the associated token, if any,
+            as an integer """
+        if self.tokentype not in ("YEAR"):
```
Comment (vthorsteinsson) [inline tree.py:781]: `if self.tokentype != "YEAR":`

File: routes/words.py (hunk 5 above)
Comment (sveinbjornt) [reply routes/words.py:97]: Good catch.

File: templates/locations/icemap.html (hunk 10 above)
Comment (sveinbjornt) [reply templates/locations/icemap.html:55]: Indeed :)

**Pull request merged**